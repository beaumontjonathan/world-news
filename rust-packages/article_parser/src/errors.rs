use chrono::ParseError as ChronoParseError;
use htmlescape::DecodeErr;
use scraper::Selector;
use serde_json::Error as SerdeJsonError;

#[derive(Debug)]
pub enum InternalParseError {
  UnknownError(&'static str),
  HtmlStructureError(HtmlStructureError),
  JsonParseError(JsonParseError),
  InvalidDateTime(ChronoParseError),
  HtmlDecodeErr(DecodeErr),
}

#[derive(Debug)]
pub enum JsonParseErrorCause {
  InvalidLdJson,
  InvalidDataAttrJson,
}

#[derive(Debug)]
pub struct JsonParseError {
  pub cause: JsonParseErrorCause,
  pub serde_error: SerdeJsonError,
}

#[derive(Debug)]
pub enum HtmlErrorCause {
  MissingElement,
  NonUniqueElement,
  MissingAttribute(String),
}

#[derive(Debug)]
pub struct HtmlStructureError {
  pub cause: HtmlErrorCause,
  pub selector: &'static Selector,
}

impl InternalParseError {
  pub fn html(error: HtmlStructureError) -> InternalParseError {
    InternalParseError::HtmlStructureError(error)
  }

  pub fn json(error: JsonParseError) -> InternalParseError {
    InternalParseError::JsonParseError(error)
  }

  pub fn ld_json(serde_error: SerdeJsonError) -> InternalParseError {
    InternalParseError::JsonParseError(JsonParseError {
      cause: JsonParseErrorCause::InvalidLdJson,
      serde_error,
    })
  }

  pub fn json_data_attr(serde_error: SerdeJsonError) -> InternalParseError {
    InternalParseError::JsonParseError(JsonParseError {
      cause: JsonParseErrorCause::InvalidDataAttrJson,
      serde_error,
    })
  }

  pub fn unknown(s: &'static str) -> InternalParseError {
    InternalParseError::UnknownError(s)
  }

  pub fn chrono(err: ChronoParseError) -> InternalParseError {
    InternalParseError::InvalidDateTime(err)
  }

  pub fn decode_html(err: DecodeErr) -> InternalParseError {
    InternalParseError::HtmlDecodeErr(err)
  }
}

impl From<ChronoParseError> for InternalParseError {
  fn from(err: ChronoParseError) -> Self {
    InternalParseError::chrono(err)
  }
}

impl From<DecodeErr> for InternalParseError {
  fn from(err: DecodeErr) -> Self {
    InternalParseError::decode_html(err)
  }
}
