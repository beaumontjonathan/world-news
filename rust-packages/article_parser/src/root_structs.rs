use chrono::{DateTime, Utc};
use serde::Serialize;
use std::fmt::Debug;

#[derive(Debug, Serialize)]
pub struct Image {
  pub alt: Option<String>,
  pub url: Option<String>,
  pub caption: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct Video {
  pub alt: Option<String>,
  pub url: Option<String>,
  pub caption: Option<String>,
}

#[derive(Clone, Debug, Serialize)]
pub enum Publisher {
  BBC,
  DailyMail,
  Independent,
  Metro,
  Mirror,
  Sky,
  Sun,
  Guardian,
}

use Publisher::*;

impl From<String> for Publisher {
  fn from(publisher: String) -> Self {
    match publisher.to_lowercase().as_str() {
      "bbc" => BBC,
      "dailymail" => DailyMail,
      "independent" => Independent,
      "metro" => Metro,
      "mirror" => Mirror,
      "sky" => Sky,
      "sun" => Sun,
      "theguardian" | "guardian" => Guardian,
      _ => panic!("Invalid publisher type"),
    }
  }
}

#[derive(Debug, Serialize)]
pub struct NewsArticleContent {
  pub headline: String,
  pub twitter_headline: Option<String>,
  pub description: String,
  pub thumbnail: Image,
  pub categories: Vec<String>,
  pub images: Vec<Image>,
  pub videos: Vec<Video>,
  pub body: String,
  pub date_updated: DateTime<Utc>,
  pub date_published: DateTime<Utc>,
}

#[derive(Debug, Serialize)]
pub struct VideoArticleContent {
  pub something: String,
}

#[allow(clippy::large_enum_variant)]
#[derive(Debug, Serialize)]
pub enum ParsedPageContent {
  NewsArticle(NewsArticleContent),
  VideoArticle(VideoArticleContent),
}

#[derive(Debug, Serialize)]
pub struct PageMeta {
  pub publisher: Publisher,
  pub url: String,
  pub date_parsed: DateTime<Utc>,
}

#[derive(Debug, Serialize)]
pub struct ParsedPage {
  pub meta: PageMeta,
  pub content: ParsedPageContent,
}
