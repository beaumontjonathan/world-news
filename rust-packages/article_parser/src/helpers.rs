use crate::errors::{HtmlErrorCause, HtmlStructureError, InternalParseError};
// use chrono::{DateTime, Utc};
use crate::root_structs::ParsedPageContent;
use lazy_static::lazy_static;
use regex::Regex;
use scraper::{ElementRef, Html, Selector};
// use std::fmt::{self, Debug, Display, Formatter};
pub trait ToOwnedOrNone {
  fn to_owned_or_none(self) -> Option<String>;
}

impl<'a> ToOwnedOrNone for Option<&'a str> {
  fn to_owned_or_none(self) -> Option<String> {
    self.and_then(|mut s| {
      s = s.trim();
      if s.is_empty() {
        None
      } else {
        Some(String::from(s))
      }
    })
  }
}
// pub trait NoneIfEmpty
// where
//   Self: Sized,
// {
//   fn none_if<F>(self, callback: F) -> Option<Self>
//   where
//     F: Fn(&Self) -> bool,
//   {
//     if callback(&self) {
//       None
//     } else {
//       Some(self)
//     }
//   }
// }
// impl<T> NoneIfEmpty for T {}

pub type ParseResult = Result<ParsedPageContent, InternalParseError>;

pub trait FixHtmlWhitespace {
  fn fix_html_whitespace(self) -> Self;
}

fn fix_whitespace(text: String) -> String {
  lazy_static! {
    static ref RE: Regex = Regex::new(r#"\s+"#).unwrap();
  }
  RE.replace_all(&text, " ").into_owned()
}

impl FixHtmlWhitespace for String {
  fn fix_html_whitespace(self) -> String {
    fix_whitespace(self)
  }
}

pub trait AttrValueOrNone {
  fn attr_value_or_none(&self, attr: &str) -> Option<String>;
}

impl<'a> AttrValueOrNone for ElementRef<'a> {
  fn attr_value_or_none(&self, attr: &str) -> Option<String> {
    self.value().attr(attr).to_owned_or_none()
  }
}

pub trait FullFixedHtmlText {
  fn full_fixed_html_text(self) -> String;
}

impl<'a> FullFixedHtmlText for ElementRef<'a> {
  fn full_fixed_html_text(self) -> String {
    self
      .text()
      .map(|text| text.trim().to_string())
      .collect::<Vec<String>>()
      .join(" ")
      .fix_html_whitespace()
      .trim()
      .to_owned()
  }
}

pub trait JoinWithNewline {
  fn join_with_newline(self) -> String;
}

impl<T> JoinWithNewline for T
where
  T: Iterator<Item = String>,
{
  fn join_with_newline(self) -> String {
    self.collect::<Vec<String>>().join("\n").trim().to_owned()
  }
}

pub trait UniqueElementRef<'a> {
  fn unique(self) -> std::result::Result<ElementRef<'a>, HtmlErrorCause>;
}

impl<'a, T> UniqueElementRef<'a> for T
where
  T: Iterator<Item = ElementRef<'a>>,
{
  fn unique(mut self) -> std::result::Result<ElementRef<'a>, HtmlErrorCause> {
    match (self.next(), self.next()) {
      (None, _) => Err(HtmlErrorCause::MissingElement),
      (_, Some(_)) => Err(HtmlErrorCause::NonUniqueElement),
      (Some(el_ref), None) => Ok(el_ref),
    }
  }
}

pub trait SelectHelpers {
  fn select_unique(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError>;

  fn select_first(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError>;

  fn select_unique_attr(
    &self,
    selector: &'static Selector,
    attr: &str,
  ) -> Result<String, InternalParseError>;
}

impl SelectHelpers for Html {
  fn select_unique(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self
      .select(selector)
      .unique()
      .map_err(|cause| InternalParseError::html(HtmlStructureError { cause, selector }))
  }

  fn select_first(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self.select(selector).next().ok_or_else(|| {
      InternalParseError::html(HtmlStructureError {
        cause: HtmlErrorCause::MissingElement,
        selector,
      })
    })
  }

  fn select_unique_attr(
    &self,
    selector: &'static Selector,
    attr: &str,
  ) -> Result<String, InternalParseError> {
    self
      .select_unique(selector)?
      .value()
      .attr(attr)
      .map(|s| s.trim().to_owned())
      .ok_or_else(|| {
        InternalParseError::html(HtmlStructureError {
          cause: HtmlErrorCause::MissingAttribute(attr.to_owned()),
          selector,
        })
      })
  }
}

impl SelectHelpers for scraper::element_ref::ElementRef<'_> {
  fn select_unique(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self
      .select(selector)
      .unique()
      .map_err(|cause| InternalParseError::html(HtmlStructureError { cause, selector }))
  }

  fn select_first(&self, selector: &'static Selector) -> Result<ElementRef, InternalParseError> {
    self.select(selector).next().ok_or_else(|| {
      InternalParseError::html(HtmlStructureError {
        cause: HtmlErrorCause::MissingElement,
        selector,
      })
    })
  }

  fn select_unique_attr(
    &self,
    selector: &'static Selector,
    attr: &str,
  ) -> Result<String, InternalParseError> {
    self
      .select_unique(selector)?
      .value()
      .attr(attr)
      .map(|s| s.trim().to_owned())
      .ok_or_else(|| {
        InternalParseError::html(HtmlStructureError {
          cause: HtmlErrorCause::MissingAttribute(attr.to_owned()),
          selector,
        })
      })
  }
}

pub trait FormattedText {
  fn formatted_text(self) -> String;
}

impl<'a, T> FormattedText for T
where
  T: Iterator<Item = ElementRef<'a>>,
{
  fn formatted_text(self) -> String {
    self.map(|p| p.full_fixed_html_text()).join_with_newline()
  }
}

pub trait ChildrenHelpers {
  fn children_first(&self) -> Option<ElementRef>;
  fn children_all(&self) -> Vec<ElementRef>;
}

impl ChildrenHelpers for scraper::element_ref::ElementRef<'_> {
  fn children_first(&self) -> Option<ElementRef> {
    self.children_all().into_iter().next()
  }

  fn children_all(&self) -> Vec<ElementRef> {
    self.children().filter_map(ElementRef::wrap).collect()
  }
}
