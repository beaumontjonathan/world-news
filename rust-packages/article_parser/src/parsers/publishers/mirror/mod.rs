use crate::errors::InternalParseError;
use crate::helpers::{
  AttrValueOrNone, FullFixedHtmlText, JoinWithNewline, ParseResult, SelectHelpers,
};
use crate::my_selectors;
use crate::root_structs::{Image, NewsArticleContent, ParsedPageContent, Video};
use chrono::DateTime;
use scraper::{Html, Selector};

my_selectors! {
  ARTICLE_ROOT_SELECTOR => "body > main > article.article-main",
  ARTICLE_CONTENT_SELECTOR => ":scope > div.article-wrapper > div.content-column",
  ARTICLE_BODY_SELECTOR => ":scope > div.article-body",
  ARTICLE_CONTENT_CHILDREN_SELECTOR => ":scope > *",
  H1_SELECTOR => "h1",
  SUB_TITLE_SELECTOR => "p.sub-title",
  THUMBNAIL_FIGURE_SELECTOR => "figure.lead-article-image",
  IMG_SELECTOR => "img",
  CAPTION_SELECTOR => "figcaption > .caption",
  URL_SELECTOR => "meta[itemprop=\"url\"]",
  ARTICLE_BODY_FIGURE_IMAGE_SELECTOR => ".article-body > figure.in-article-image",
  ARTICLE_BODY_P_AND_URL_SELECTOR => ":scope > p, :scope > ul",
  DATE_UPDATED_SELECTOR => "head > meta[property=\"article:modified_time\"]",
  DATE_PUBLISHED_SELECTOR => "head > meta[property=\"article:published_time\"]",
  DIV_MOD_VIDEO => "div.mod-video",
  FIGURE_LEAD_ARTICLE_IMAGE => "figure.lead-article-image",
  FIGURE_IN_ARTICLE_IMAGE => ":scope > figure.in-article-image",
}

pub fn parse(document: Html) -> ParseResult {
  // DOCUMENT SELECTED
  let article_root = document.select_unique(&ARTICLE_ROOT_SELECTOR)?;
  let article_content = article_root.select_unique(&ARTICLE_CONTENT_SELECTOR)?;
  let article_body = article_content.select_unique(&ARTICLE_BODY_SELECTOR)?;

  // HEADLINE
  let headline = article_root
    .select_unique(&H1_SELECTOR)?
    .full_fixed_html_text();

  // TWITTER HEADLINE
  let twitter_headline = None;

  // DESCRIPTION
  let description: String = article_root
    .select_unique(&SUB_TITLE_SELECTOR)
    .map(|sub_title| sub_title.full_fixed_html_text())
    .unwrap_or_else(|_| String::new());

  // THUMBNAIL
  let thumbnail: Image = {
    let first_child = article_content.select_first(&ARTICLE_CONTENT_CHILDREN_SELECTOR)?;
    if DIV_MOD_VIDEO.matches(&first_child) {
      // todo - could get data from the "div.json-placeholder" element, but
      // html escaping breaks the json - e.g., double quotes in headline text.
      Ok(Image {
        alt: None,
        url: None,
        caption: None,
      })
    } else if FIGURE_LEAD_ARTICLE_IMAGE.matches(&first_child) {
      let caption = Some(
        first_child
          .select_unique(&CAPTION_SELECTOR)?
          .full_fixed_html_text(),
      );
      let img = first_child.select_unique(&IMG_SELECTOR)?;
      let url = img.attr_value_or_none("src");
      let alt = img.attr_value_or_none("alt");
      Ok(Image { alt, url, caption })
    } else {
      Err(InternalParseError::unknown("thumbnail type unknown"))
    }
  }?;

  // CATEGORIES
  // TODO - implement
  let categories: Vec<String> = vec![];

  // IMAGES
  let images = article_body
    .select(&FIGURE_IN_ARTICLE_IMAGE)
    .map(|figure| {
      let img = figure.select_first(&IMG_SELECTOR)?;
      let alt = img.value().attr("alt").and_then(|s| match s.len() {
        0 => None,
        _ => Some(s.to_owned()),
      });
      let url = figure
        .select_unique(&URL_SELECTOR)?
        .attr_value_or_none("content");
      let caption = figure
        .select(&CAPTION_SELECTOR)
        .next()
        .map(|el| el.full_fixed_html_text());
      Ok(Image { url, alt, caption })
    })
    .collect::<Result<_, InternalParseError>>()?;

  // VIDEOS
  // todo - implement
  let videos: Vec<Video> = vec![];

  // BODY
  let body = article_body
    .select(&ARTICLE_BODY_P_AND_URL_SELECTOR)
    .map(|el| el.full_fixed_html_text())
    .join_with_newline();

  // DATE UPDATED
  let date_updated = document.select_unique_attr(&DATE_UPDATED_SELECTOR, "content")?;
  let date_updated = DateTime::parse_from_rfc3339(&date_updated)?.into();

  //DATE PUBLISHED
  let date_published = document.select_unique_attr(&DATE_PUBLISHED_SELECTOR, "content")?;
  let date_published = DateTime::parse_from_rfc3339(&date_published)?.into();

  Ok(ParsedPageContent::NewsArticle(NewsArticleContent {
    headline,
    twitter_headline,
    description,
    thumbnail,
    categories,
    images,
    videos,
    body,
    date_updated,
    date_published,
  }))
}

#[cfg(test)]
mod test {
  use super::*;
  use crate::test_helpers::helpers::read_html_doc;
  use insta::assert_debug_snapshot;

  #[test]
  fn test_file_01() {
    let document = read_html_doc("test-01.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_02() {
    let document = read_html_doc("test-02.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_03() {
    let document = read_html_doc("test-03.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_04() {
    let document = read_html_doc("test-04.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_06() {
    let document = read_html_doc("test-06.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }

  #[test]
  fn test_file_07() {
    let document = read_html_doc("test-07.html", std::file!());
    let article = parse(document);
    assert_debug_snapshot!(article);
  }
}
