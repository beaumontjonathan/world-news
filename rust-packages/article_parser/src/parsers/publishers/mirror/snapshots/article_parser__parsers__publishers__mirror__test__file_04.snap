---
source: src/parsers/publishers/mirror/mod.rs
expression: article
---
Ok(
    NewsArticle(
        NewsArticleContent {
            headline: "Queen whisked away from Buckingham Palace as coronavirus quarantine plans explained",
            twitter_headline: None,
            description: "The Queen and Prince Philip will be put in quarantine away from Buckingham Palace if the crippling coronavirus (Covid-19) pandemic worsens",
            thumbnail: Image {
                alt: None,
                url: Some(
                    "https://i2-prod.mirror.co.uk/incoming/article21269262.ece/ALTERNATES/s615/2_BESTPIX-Queen-Attends-Sunday-Church-in-Sandringham.jpg",
                ),
                caption: Some(
                    "The Queen is pictured yesterday leaving church recently",
                ),
            },
            categories: [],
            images: [
                Image {
                    alt: None,
                    url: Some(
                        "https://i2-prod.mirror.co.uk/incoming/article21661878.ece/ALTERNATES/s615b/0_Commonwealth-Day-Service-2020.jpg",
                    ),
                    caption: Some(
                        "Queen Elizabeth II attends the Commonwealth Day Service",
                    ),
                },
                Image {
                    alt: None,
                    url: Some(
                        "https://i2-prod.mirror.co.uk/incoming/article21654273.ece/ALTERNATES/s615b/1_Queen-Elizabeth-II-attends-church-at-Windsor-UK-08-Mar-2020.jpg",
                    ),
                    caption: Some(
                        "Queen Elizabeth II leaves Windsor Castle to attend a function",
                    ),
                },
            ],
            videos: [],
            body: "The Queen has been chauffeured away from Buckingham Palace amid the coronavirus outbreak.\nIt is said the 93-year-old monarch was whisked to Windsor Castle on Thursday for her safety.\nBut a source at Buckingham Palace told Mirror Online the Queen went to Windsor on her usual weekend trip and will be back in London next week.\nBut it is understood plans are now in place to quarantine the Queen and Prince Philip, 98, at Sandringham if the outbreak worsens.\nHowever, there had been reports the Queen had been \"whisked away\" from the Palace over cronavirus fears.\nA royal source said: \"She is in good health but it was thought best to move her. A lot of her staff are a bit panicky over coronavirus.\n“The Palace hosts a constant stream of visitors including politicians and dignitaries from around the world.\"\nThe Sun reports advisers have suggested the move for the Queen as the death toll to Covid-19 rises in the UK.\nMore than 20 have succumbed to the virus in the country, with confirmed cases rocketing by 342 to 1,140.\nThe royal source added: \"The Queen has met a lot of people there until recently. But she is weeks away from her 94th birthday and advisers believe it is best to get her out of harm’s way.\n“Buckingham Palace is in the middle of London and also has a bigger staff than other estates so is deemed a much more dangerous location.\n“There have been no specific scares or positive tests there yet but no one wants to take any chances.”\nThe Palace has around 500 staff, Windsor 100 and Sandringham a handful.\nPalace garden parties for 30,000 guests in May and June may be cancelled because of the pandemic.",
            date_updated: 2020-03-14T23:59:32Z,
            date_published: 2020-03-14T22:39:08Z,
        },
    ),
)
